server/src/main/java/org/elasticsearch/watcher/ResourceWatcherService.java

```diff
@@ -78,51 +77,42 @@
         Setting.timeSetting("resource.reload.interval.low", Frequency.LOW.interval, Property.NodeScope);
 
     private final boolean enabled;
-    private final ThreadPool threadPool;
 
     final ResourceMonitor lowMonitor;
     final ResourceMonitor mediumMonitor;
     final ResourceMonitor highMonitor;
 
-    private volatile Cancellable lowFuture;
-    private volatile Cancellable mediumFuture;
-    private volatile Cancellable highFuture;
+    private final Cancellable lowFuture;
+    private final Cancellable mediumFuture;
+    private final Cancellable highFuture;
 
-    @Inject
     public ResourceWatcherService(Settings settings, ThreadPool threadPool) {
         this.enabled = ENABLED.get(settings);
-        this.threadPool = threadPool;
 
         TimeValue interval = RELOAD_INTERVAL_LOW.get(settings);
         lowMonitor = new ResourceMonitor(interval, Frequency.LOW);
         interval = RELOAD_INTERVAL_MEDIUM.get(settings);
         mediumMonitor = new ResourceMonitor(interval, Frequency.MEDIUM);
         interval = RELOAD_INTERVAL_HIGH.get(settings);
         highMonitor = new ResourceMonitor(interval, Frequency.HIGH);
-    }
-
-    @Override
-    protected void doStart() {
-        if (!enabled) {
-            return;
+        if (enabled) {
+            lowFuture = threadPool.scheduleWithFixedDelay(lowMonitor, lowMonitor.interval, Names.SAME);
+            mediumFuture = threadPool.scheduleWithFixedDelay(mediumMonitor, mediumMonitor.interval, Names.SAME);
+            highFuture = threadPool.scheduleWithFixedDelay(highMonitor, highMonitor.interval, Names.SAME);
```

Comment:
This schedule will still begin its first poll after the initial delay, which is the "big" gap for #54867. Or theoretically, anything added between 2 polls may have a chance to miss an update. You are probably solving it in #54999 (I haven't looked through). Even so, I wonder whether it would be better if this watch service can guarantee no gaps.

Reply:
Correct, there will still be the initial delay but we won't miss an update AFAICS. The `add` method initializes the watcher that is being registered: https://github.com/elastic/elasticsearch/blob/af7ea67541bbe8d2aa4f7e6ca53ddae942276c74/server/src/main/java/org/elasticsearch/watcher/ResourceWatcherService.java#L138-L139

Given that we are monitoring files, this will be a `FileWatcher` which extends `AbstractResourceWatcher`. `AbstractResourceWatcher` has an `init` method that will call the `doInit` method in the `FileWatcher` class https://github.com/elastic/elasticsearch/blob/af7ea67541bbe8d2aa4f7e6ca53ddae942276c74/server/src/main/java/org/elasticsearch/watcher/AbstractResourceWatcher.java#L33-L39

The `FileWatcher#doInit` method will initialize the `FileObserver` https://github.com/elastic/elasticsearch/blob/af7ea67541bbe8d2aa4f7e6ca53ddae942276c74/server/src/main/java/org/elasticsearch/watcher/FileWatcher.java#L63-L66

The `FileObserver` initialization captures whether the directory we are monitoring exists  https://github.com/elastic/elasticsearch/blob/af7ea67541bbe8d2aa4f7e6ca53ddae942276c74/server/src/main/java/org/elasticsearch/watcher/FileWatcher.java#L156-L169

This will then trigger the listing of the child files within that directory and sub directories https://github.com/elastic/elasticsearch/blob/af7ea67541bbe8d2aa4f7e6ca53ddae942276c74/server/src/main/java/org/elasticsearch/watcher/FileWatcher.java#L287-L300

A `FileObserver` is created for each child https://github.com/elastic/elasticsearch/blob/af7ea67541bbe8d2aa4f7e6ca53ddae942276c74/server/src/main/java/org/elasticsearch/watcher/FileWatcher.java#L183-L194

Each child is also initialized https://github.com/elastic/elasticsearch/blob/af7ea67541bbe8d2aa4f7e6ca53ddae942276c74/server/src/main/java/org/elasticsearch/watcher/FileWatcher.java#L171-L175

This means that we will not miss any updates to these files if we register the monitoring of the files prior to actually loading the files, which is where #54999 comes into play. This change will allow for changes to possibly be found sooner.

Reply:
You are absolutely right. I completely missed that `watcher.init();` call. 🤦